edb/schema/links.py

```diff
@@ -100,20 +106,27 @@ class Link(sources.Source, pointers.Pointer, s_abc.Link,
         compcoef=0.9,
         merge_fn=merge_actions)
 
-    def is_link_property(self, schema):
+    def is_link_property(self, schema: s_schema.Schema) -> bool:
         return False
 
-    def is_property(self, schema):
+    def is_property(self, schema: s_schema.Schema) -> bool:
         return False
 
-    def scalar(self):
+    def scalar(self) -> bool:
         return False
 
-    def has_user_defined_properties(self, schema):
+    def has_user_defined_properties(self, schema: s_schema.Schema) -> bool:
         return bool([p for p in self.get_pointers(schema).objects(schema)
                      if not p.is_special_pointer(schema)])
 
-    def compare(self, other, *, our_schema, their_schema, context=None):
+    def compare(
+        self,
+        other: so.Object,
+        *,
+        our_schema: s_schema.Schema,
+        their_schema: s_schema.Schema,
+        context: Optional[so.ComparisonContext] = None,
+    ) -> Union[float, NotImplemented]:
```

Comment:
In my pending branch I replaced returning `NotImplemented` with `raise NotImplementedError`, please do the same here.